plugin/plugin.go

```diff
@@ -0,0 +1,99 @@
+package plugin
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	"github.com/tetratelabs/wazero"
+)
+
+const (
+	PluginName = "wasm"
+)
+
+// New initializes a new plugin and returns it.
+func New(guestPath string /*runtime.Object, framework.Handle*/) (framework.Plugin, error) {
+	// TODO: make this configuration via URL
+	const guestName = "example"
+
+	ctx := context.Background()
+
+	runtime, guestModule, err := prepareRuntime(ctx, guestPath)
+	if err != nil {
+		return nil, err
+	}
+
+	pl := &wasmPlugin{
+		guestModuleConfig: wazero.NewModuleConfig(),
+		guestName:         guestName,
+		runtime:           runtime,
+		guestModule:       guestModule,
+		instanceCounter:   atomic.Uint64{},
+	}
+
+	// Eagerly add one instance to the pool. Doing so helps to fail fast.
+	g, err := pl.getOrCreateGuest(ctx)
+	if err != nil {
+		_ = runtime.Close(ctx)
+		return nil, err
+	}
+	pl.pool.Put(g)
```

Comment:
Cannot we initialize one guest here and use the same guest instance in all functions? What's the reason of the need of pool?

Reply:
my assumption is that there are multiple parallel requests to the scheduler. If not, we can indeed make it a single guest.

Reply:
Got it, thanks!
Yes, each plugin can receive multiple parallel requests. So, I think that makes sense.

Reply:
fyi @pohly

Reply:
> no wasm guest is safe for concurrent use

Is that a limitation of the current runtime and may that get addressed at some point?

Reply:
TL;DR; I think at least for a year maybe two, assume all guests have to be used sequentially.

Even if the wasm runtime supports atomics, the guest would have to be compiled to use them. One issue is garbage collector implementations (compiled to wasm) would need to use them. It isn't just that, as stack pointers etc would also need to be changed. I don't expect TinyGo or Go to do that for a long time. Even though there's a GC proposal in wasm which may affect some of this, it wasn't written with Go in mind. For example, it doesn't support interior pointers. I don't think any garbage collected language compiled to wasm, or even a malloc/free language is using atomics yet.

Reply:
> Yes, each plugin can receive multiple parallel requests

Can you elaborate? Why would we receive parallel requests?

Reply:
Regarding the scheduling cycle, yes. But we have a binding cycle running in parallel.

Reply:
Also, we run Filter plugins in default preemption as well
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/preemption/preemption.go#LL606C77-L606C77